### ISSUE TITLE
Support for execution flow ports

### ISSUE THREAD

**Issue by stonelance** [Wouterdek/NodeNetwork#2]
First of all, thanks for this library.  It seems like a great start and I've had a hard time finding anything else similar before.

I'm not sure if this is outside of the intended use cases for this library, but I am investigating doing some codegen using a node network, and would like to have "execution flow" ports in addition to the "data flow" ports.  The main distinction is that they are evaluated in the opposite order, and multiple connections on an output is invalid, but multiple connections on an input is valid.  This is similar to functionality in Unreal Engine's blueprint system if you are familiar with it.

Is this something that is reasonable to support in this library?  If I wanted to go about adding support for it, what do you think the best approach would be?  I tried actually just using ValueNodeOutputViewModel for inputs and ValueNodeInputViewModel for outputs, as this is essentially what I need but it doesn't work because NodeViewModel expects inputs to be derived from NodeInputViewModel and outputs derived from NodeOutputViewModel.  I'm guessing I would run into other issues with that approach anyway :)

**Comment by Wouterdek** [comment]
I am not familiar with the blueprint system, but at first glance it looks like something that should be possible with this library. I'll download Unreal tonight, try it out for a bit and see what the best approach is.

**Comment by stonelance** [comment]
That would be awesome, thanks.  There are some plenty of good tutorial videos on it.  Here is a fairly simple example with a for loop that helps demonstrate the execution flow: https://www.youtube.com/watch?v=hXSKGYQvZsI

**Comment by Wouterdek** [comment]
Ok, so I have played around with Blueprint and I'm sure NodeNetwork can support something similar if a few changes are made. Some inputs in Blueprint support multiple connections (ex. execution inputs, 'target' inputs on some nodes, ...). NodeNetwork does not support that currently, but that should be easy to fix.

Once those changes are made, implementing a Blueprint-like system should work something like this:

- Blueprint has red/green/blue/white/... nodes. This can be done in NodeNetwork by creating a custom node view that supports different colors based on the node viewmodel type.
- Blueprint has ports with different shapes/colors based on the type of input/output. This can be done by creating a custom port view that changes shape/color depending on the input/output type.
- In Blueprint data inputs/outputs are positioned left/right like in NodeNetwork. However, execution flow inputs/outputs are right/left. This can be reproduced in NodeNetwork by creating a custom Input/Output view that swaps the placement of the ports.
- Blueprint has checkboxes/textboxes/... next to their inputs that provide a fallback value. This is supported in NodeNetwork through the 'Editor' property on inputs. Both the calculator and the shader editor have examples of how to do this.
- In Blueprint, the color of the connections between nodes is based on the type of input/output. This can be done in NodeNetwork by creating a custom connection view which does the same thing.
- To generate the code, you could use a similar system used in the shader editor example. Find the starting nodes for the execution flows and recursively compile the function produced by each node.

This may seem like a lot of work, but once you understand how it works these are fairly simple to do. Also, you may not need/want all of the things I listed.

I will implement something like this and add it to the examples in the project, but note that this could take a while as my time is currently fairly limited.

**Comment by stonelance** [comment]
Ok thanks for taking a look at it.  I'll play around with some different ideas on my end as well.  Is there a way to serialize\deserialize the graphs?

**Comment by Wouterdek** [comment]
If you want to serialize the viewmodels, that is on my TODO list. It should be fairly easy since there is clear separation between viewmodel and view and so in theory the viewmodel classes should be serializable. However, it's possible that some .net serialization properties are needed here and there to make everything work correctly.

Or maybe you want to convert the graph from the viewmodel to your own model class? If so, I think it would be easiest to include a `Build()` function in your node subclasses and then loop over `NetworkViewModel.Nodes` and build each one to get the corresponding model object.

**Comment by Wouterdek** [comment]
I have made several commits that change some aspects of the library and add a new example project: ExampleCodeGenApp

Its currently only available in source code, but try it out and see if it fits your needs. It should function similarly to Blueprint.

**Comment by stonelance** [comment]
Oh wow, thanks  I'll take a look.

I haven't worked on this in a week or two now, but I had followed your advice and made some changes to the NodeNetwork code to get this setup working: https://youtu.be/oq_6wEfKd1w  I ended up avoiding the execution flow port issue by not using the graph evaluation functionality in the NodeNetwork itself and instead converting to my own model representation of the graph, and then running my codegen from that.  It also allowed me to implement serialization from that model, and then I recreate the viewmodel on load.  There are still some downsides of that, like i was still limited to one connection to an input port.  I'll look at merging with your latest and see if there is anything I have that might be worth pulling back into NodeNetwork.
